File: Universal.Xaml.Analytics/AnalyticsPlugin.cs
using Microsoft.Media.Analytics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
#if SILVERLIGHT
using System.Windows;
using System.Windows.Threading;
#else
using System.Threading.Tasks;
using Windows.UI.Xaml;
using Windows.UI.Core;
#endif

namespace Microsoft.PlayerFramework.Analytics
{
    /// <summary>
    /// A plugin used to retrieve and log analytics data.
    /// </summary>
    public class AnalyticsPlugin : IPlugin
    {
        /// <summary>
        /// Gets the area key to be applied to all analytic related tracking events.
        /// Tracking events without this area will be ignored by the analytics plugin.
        /// </summary>
        public static string TrackingEventArea { get { return "Analytics"; } }

        /// <summary>
        /// Creates a new instance of AnalyticsPlugin.
        /// </summary>
        public AnalyticsPlugin()
        {
            AnalyticsConfig = new AnalyticsConfig();
            MediaData = new Dictionary<string, object>();
            SessionData = new Dictionary<string, object>();
        }

        /// <summary>
        /// Creates a new instance of AnalyticsPlugin.
        /// </summary>
        /// <param name="analyticsConfig">The required analytics config object used to control what kind of analytics data should be collected.</param>
        public AnalyticsPlugin(AnalyticsConfig analyticsConfig)
        {
            AnalyticsConfig = analyticsConfig ?? new AnalyticsConfig();
            MediaData = new Dictionary<string, object>();
            SessionData = new Dictionary<string, object>();
        }

        AnalyticsCollector collector;
        MediaPlayerAdapter playerMonitor;

        /// <summary>
        /// Gets or sets the a Dispatcher to be used to poll CPU and FPS metrics. Set to null if one is not needed.
        /// </summary>
#if SILVERLIGHT
        public Dispatcher Dispatcher { get; set; }
#else
        public CoreDispatcher Dispatcher { get; set; }
#endif

        /// <summary>
        /// Gets or sets an object responsible for providing environment or system level data to log.
        /// </summary>
        public IEnvironmentMonitor EnvironmentMonitor { get; set; }

        /// <summary>
        /// Gets or sets an object responsible for providing information about the edge server used to serve the media.
        /// </summary>
        public IEdgeServerMonitor EdgeServerMonitor { get; set; }

        /// <summary>
        /// Gets or sets an object responsible for providing additional information about adaptive streaming.
        /// </summary>
        public IAdaptiveMonitor AdaptiveMonitor { get; set; }

        /// <summary>
        /// Gets or sets the required configuration object to help determine what information to collect, aggregate and log.
        /// </summary>
        public AnalyticsConfig AnalyticsConfig { get; set; }

        /// <summary>
        /// Gets the analytics collector object used by the analytics plugin to collect data
        /// </summary>
        public AnalyticsCollector AnalyticsCollector { get { return collector; } }

        /// <summary>
        /// Gets a collection of data to add to each log only for the current media. (e.g. 'VideoId')
        /// </summary>
        public IDictionary<string, object> MediaData { get; private set; }

        /// <summary>
        /// Gets a collection of data to add to each log across all media. (e.g. 'AppId')
        /// </summary>
        public IDictionary<string, object> SessionData { get; private set; }

        private void AddAdditionalData(IDictionary<string, object> data)
        {
            if (data != null)
            {
                foreach (var item in data)
                {
                    collector.AddtionalData.Add(item);
                }
            }
        }

        private void RemoveAdditionalData(IDictionary<string, object> data)
        {
            if (data != null)
            {
                foreach (var item in data)
                {
                    collector.AddtionalData.Remove(item);
                }
            }
        }

        /// <summary>
        /// Sends a log to the LoggingService but stamps it with session and media specific data first.
        /// </summary>
        /// <param name="log">The log to send</param>
        public void Log(ILog log)
        {
            collector.SendLog(log);
        }

        /// <inheritdoc /> 
        public void Load()
        {
            // by default, we always add the collector as a logging source
            collector = new AnalyticsCollector(AnalyticsConfig);
            collector.Dispatcher = Dispatcher;
            // add session specific data
            AddAdditionalData(SessionData);
            LoggingService.Current.LoggingSources.Add(collector);

            // initialize the collector. The AnalyticsCollector relies on other objects to pass it info.
            playerMonitor = new MediaPlayerAdapter(MediaPlayer);
            AddAdditionalData(MediaData);
            collector.Attach(playerMonitor, AdaptiveMonitor, EnvironmentMonitor, EdgeServerMonitor);
        }

        /// <inheritdoc /> 
        public void Update(IMediaSource mediaSource)
        {
            if (collector.IsAttached)
            {
                collector.Detach();
                // remove media specific data
                RemoveAdditionalData(MediaData);
                MediaData = null;
            }

            if (mediaSource != null)
            {
                // add media specific data
                MediaData = Analytics.GetAdditionalData(mediaSource as DependencyObject);
                AddAdditionalData(MediaData);
                collector.Attach(playerMonitor, AdaptiveMonitor, EnvironmentMonitor, EdgeServerMonitor);
            }
        }

        /// <inheritdoc /> 
        public void Unload()
        {
            collector.Detach();
            // remove session specific data
            RemoveAdditionalData(SessionData);
            collector.Dispatcher = null;
            collector = null;
            playerMonitor.MediaPlayer = null;
            playerMonitor = null;
        }

        /// <inheritdoc /> 
        public MediaPlayer MediaPlayer { get; set; }
    }
}

File: Universal.Js.Analytics/js/AnalyticsPlugin.js
(function (PlayerFramework, undefined) {
    "use strict";

    // AnalyticsPlugin Errors
    var invalidConstruction = "Invalid construction: AnalyticsPlugin constructor must be called using the \"new\" operator.";

    // AnalyticsPlugin Class
    var AnalyticsPlugin = WinJS.Class.derive(PlayerFramework.PluginBase, function (options) {
        if (!(this instanceof PlayerFramework.Plugins.AnalyticsPlugin)) {
            throw invalidConstruction;
        }

        this._collector = null;
        this._playerMonitor = null;
        this._environmentMonitor = null;
        this._edgeServerMonitor = null;
        this._adaptiveMonitor = null;
        this._analyticsConfig = null;
        this._sessionData = [];
        this._mediaData = [];
        this._mediaPlayerAdapter = null;

        PlayerFramework.PluginBase.call(this, options);
    }, {
        // Public Properties
        environmentMonitor: {
            get: function () {
                return this._environmentMonitor;
            },
            set: function (value) {
                this._environmentMonitor = value;
            }
        },

        edgeServerMonitor: {
            get: function () {
                return this._edgeServerMonitor;
            },
            set: function (value) {
                this._edgeServerMonitor = value;
            }
        },

        adaptiveMonitor: {
            get: function () {
                return this._adaptiveMonitor;
            },
            set: function (value) {
                this._adaptiveMonitor = value;
            }
        },

        analyticsConfig: {
            get: function () {
                return this._analyticsConfig;
            },
            set: function (value) {
                this._analyticsConfig = value;
            }
        },
        
        analyticsCollector: {
            get: function () {
                return this._collector;
            }
        },

        sessionData: {
            get: function () {
                return this._sessionData;
            },
            set: function (value) {
                this._sessionData = value;
            }
        },

        mediaData: {
            get: function () {
                return this._mediaData;
            },
            set: function (value) {
                this._mediaData = value;
            }
        },

        // Private Methods
        _setOptions: function (options) {
            PlayerFramework.Utilities.setOptions(this, options, {
                isEnabled: true,
                analyticsConfig: new Microsoft.Media.Analytics.AnalyticsConfig()
            });
        },

        _onActivate: function () {
            // by default, we always add the AnalyticsCollector as a logging source
            this._collector = new Microsoft.Media.Analytics.AnalyticsCollector();
            // add session specific data
            this._addAdditionalData(this.sessionData);
            Microsoft.Media.Analytics.LoggingService.current.loggingSources.append(this._collector);

            // initialize the AnalyticsCollector. The analytics collector relies on other objects to pass it info.
            this._mediaPlayerAdapter = new PlayerFramework.Analytics.MediaPlayerAdapter(this.mediaPlayer);
            this._playerMonitor = this._mediaPlayerAdapter.nativeInstance;

            this._bindEvent("loadstart", this.mediaPlayer, this._onMediaPlayerLoadStart);
            this._bindEvent("emptied", this.mediaPlayer, this._onMediaPlayerEmptied);

            return true;
        },

        _onDeactivate: function () {
            if (this._collector.isAttached) {
                this._collector.detach();
            }
            this._unbindEvent("loadstart", this.mediaPlayer, this._onMediaPlayerLoadStart);
            this._unbindEvent("emptied", this.mediaPlayer, this._onMediaPlayerEmptied);
            // remove session specific data
            this._removeAdditionalData(this.sessionData);
            this._collector = null;
            this._mediaPlayerAdapter.dispose();
            this._mediaPlayerAdapter = null;
            this._playerMonitor = null;
        },

        _addAdditionalData: function (data) {
            if (data) {
                for (var i = 0; i < data.length; i++) {
                    var item = data[i];
                    this._collector.addtionalData.insert(item.key, item.value);
                }
            }
        },

        _removeAdditionalData: function (data) {
            if (data) {
                for (var i = 0; i < data.length; i++) {
                    var item = data[i];
                    this._collector.addtionalData.remove(item.key);
                }
            }
        },
        
        _onMediaPlayerLoadStart: function () {
            this._collector.configuration = this.analyticsConfig;

            // attach the AnalyticsCollector
            if (!this._collector.isAttached) {
                this._collector.attach(this._playerMonitor, this.adaptiveMonitor, this.environmentMonitor, this.edgeServerMonitor);
            }
            // remove media specific data
            this._addAdditionalData(this.mediaData);
        },
        
        _onMediaPlayerEmptied: function () {
            // remove media specific data
            this._removeAdditionalData(this.mediaData);
            // detach the AnalyticsCollector
            if (this._collector.isAttached) {
                this._collector.detach();
            }

            this._collector.configuration = null;
        },

        // Public Methods
        log: function (log) {
            this._collector.sendLog(log);
        }
    });

    AnalyticsPlugin.trackingEventArea = "analytics";

    // AnalyticsPlugin Mixins
    WinJS.Class.mix(PlayerFramework.MediaPlayer, {
        AnalyticsPlugin: {
            value: null,
            writable: true,
            configurable: true
        }
    });

    // AnalyticsPlugin Exports
    WinJS.Namespace.define("PlayerFramework.Plugins", {
        AnalyticsPlugin: AnalyticsPlugin
    });

})(PlayerFramework);

